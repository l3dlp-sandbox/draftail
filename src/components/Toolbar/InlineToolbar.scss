.#{$DRAFTAIL}ToolbarButton--pin {
  color: $draftail-editor-chrome;
  background-color: $draftail-editor-chrome-text;
  border-end-start-radius: $draftail-toolbar-radius;
  padding: 0;
  min-width: 1.5rem;
  height: 1.5rem;

  .#{$DRAFTAIL}Icon {
    width: 0.75rem;
    height: 0.75rem;
  }

  .#{$DRAFTAIL}ToolbarButton__label {
    vertical-align: top;
  }

  &::before {
    margin-bottom: 2px;
  }

  &::after {
    margin-bottom: 8px;
  }

  .#{$DRAFTAIL}FloatingToolbar & {
    vertical-align: top;
    transform: translate($controls-spacing, -1 * $controls-spacing);
  }

  .#{$DRAFTAIL}Toolbar & {
    position: absolute;
    top: 1px;
    inset-inline-end: 1px;
  }
}
