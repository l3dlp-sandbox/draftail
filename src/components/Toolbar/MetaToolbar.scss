.#{$DRAFTAIL}MetaToolbar {
  position: relative;
  display: flex;
  justify-content: flex-end;
  padding: $controls-spacing;
  background-color: $draftail-editor-background;
  color: $draftail-editor-chrome;
  border-top: 1px solid transparent;
}
