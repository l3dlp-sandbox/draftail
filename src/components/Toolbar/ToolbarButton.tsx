import React, { PureComponent } from "react";
import { IconProp } from "../../api/types";
import Icon from "../Icon";

export interface ToolbarButtonProps {
  name?: string;
  active?: boolean;
  label?: string | null;
  title?: string | null;
  icon?: IconProp | null;
  onClick?: ((name: string) => void) | null;
}

interface ToolbarButtonState {
  showTooltipOnHover: boolean;
}
/**
 * Displays a basic button, with optional active variant,
 * enriched with a tooltip. The tooltip stops showing on click.
 */
class ToolbarButton extends PureComponent<
  ToolbarButtonProps,
  ToolbarButtonState
> {
  constructor(props: ToolbarButtonProps) {
    super(props);

    this.state = {
      showTooltipOnHover: true,
    };

    this.onMouseDown = this.onMouseDown.bind(this);
    this.onMouseLeave = this.onMouseLeave.bind(this);
  }

  onMouseDown(e: React.MouseEvent<HTMLButtonElement>) {
    const { name, onClick } = this.props;

    e.preventDefault();

    this.setState({
      showTooltipOnHover: false,
    });

    if (onClick) {
      onClick(name || "");
    }
  }

  onMouseLeave() {
    this.setState({
      showTooltipOnHover: true,
    });
  }

  render() {
    const { name, active, label, title, icon } = this.props;
    const { showTooltipOnHover } = this.state;

    return (
      <button
        name={name}
        className={`Draftail-ToolbarButton${
          active ? " Draftail-ToolbarButton--active" : ""
        }`}
        type="button"
        aria-label={title || undefined}
        data-draftail-balloon={title && showTooltipOnHover ? true : null}
        tabIndex={-1}
        onMouseDown={this.onMouseDown}
        onMouseLeave={this.onMouseLeave}
      >
        {icon ? <Icon icon={icon} /> : null}
        {label ? (
          <span className="Draftail-ToolbarButton__label">{label}</span>
        ) : null}
      </button>
    );
  }
}

export default ToolbarButton;
