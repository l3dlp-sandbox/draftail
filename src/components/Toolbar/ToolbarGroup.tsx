import React from "react";

interface ToolbarGroupProps {
  children?: React.ReactNode;
}

const ToolbarGroup = ({ children }: ToolbarGroupProps) => {
  const hasChildren = React.Children.toArray(children).some((c) => c !== null);
  return hasChildren ? (
    <div className="Draftail-ToolbarGroup">{children}</div>
  ) : null;
};

export default ToolbarGroup;
