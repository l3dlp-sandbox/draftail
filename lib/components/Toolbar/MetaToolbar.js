// @flow
import React from "react";
import type { ComponentType } from "react";
import { EditorState } from "draft-js";

import type { IconProp } from "../Icon";
import ToolbarButton from "./ToolbarButton";
import ToolbarGroup from "./ToolbarGroup";
import { showButton, getButtonLabel, getButtonTitle } from "./ToolbarDefaults";

type ControlProps = {|
  getEditorState: () => EditorState,
  onChange: (EditorState) => void,
|};
type ControlComponent = ComponentType<ControlProps>;
type LegacyControlConfig = ControlComponent;
type CurrentControlConfig = {|
  inline?: ControlComponent,
  block?: ControlComponent,
  meta?: ControlComponent,
|};
type ControlConfig = CurrentControlConfig;

export type MetaToolbarProps = {
  showBlockEntities: ?boolean,
  entityTypes: $ReadOnlyArray<{
    type: string,
    // Describes the control in the editor UI, concisely.
    label?: ?string,
    // Describes the control in the editor UI.
    description?: string,
    // Represents the control in the editor UI.
    icon?: IconProp,
    block?: ComponentType<{}>,
  }>,
  controls: $ReadOnlyArray<ControlConfig>,
  getEditorState: () => EditorState,
  onChange: (EditorState) => void,
  onRequestSource: (entityType: string) => void,
};

const MetaToolbar = ({
  showBlockEntities,
  entityTypes,
  controls,
  getEditorState,
  onChange,
  onRequestSource,
}: MetaToolbarProps) => (
  <div className="Draftail-MetaToolbar">
    {showBlockEntities ? (
      <ToolbarGroup key="entities">
        {entityTypes
          .filter((entityType) => showButton(entityType) && entityType.block)
          .map((t) => (
            <ToolbarButton
              key={t.type}
              name={t.type}
              onClick={onRequestSource}
              label={getButtonLabel(t.type, t)}
              title={getButtonTitle(t.type, t)}
              icon={t.icon}
            />
          ))}
      </ToolbarGroup>
    ) : null}
    <ToolbarGroup>
      {controls.map((control, i) => {
        if (control.inline || control.block) {
          return null;
        }

        // Support the legacy and current controls APIs.
        // $FlowFixMe
        const Control = control.meta || (control: LegacyControlConfig);

        return (
          <Control
            // eslint-disable-next-line @thibaudcolas/cookbook/react/no-array-index-key
            key={i}
            getEditorState={getEditorState}
            onChange={onChange}
          />
        );
      })}
    </ToolbarGroup>
  </div>
);

export default MetaToolbar;
