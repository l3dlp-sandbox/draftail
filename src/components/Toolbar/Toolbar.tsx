import React from "react";
import type { ComponentType } from "react";
import { EditorState } from "draft-js";

import ToolbarDefaults, { ToolbarDefaultProps } from "./ToolbarDefaults";
import ToolbarGroup from "./ToolbarGroup";

type ControlProps = {
  getEditorState: () => EditorState;
  onChange: (arg0: EditorState) => void;
};

type ControlComponent = ComponentType<ControlProps>;
type LegacyControlConfig = ControlComponent;
type CurrentControlConfig = {
  inline?: ControlComponent;
  block?: ControlComponent;
  meta?: ControlComponent;
};

type ControlConfig = CurrentControlConfig | LegacyControlConfig;
export interface ToolbarProps extends ToolbarDefaultProps {
  controls: ReadonlyArray<ControlConfig>;
  getEditorState: () => EditorState;
  onChange: (arg0: EditorState) => void;
}

const Toolbar = (props: ToolbarProps) => {
  // Support the legacy and current controls APIs.
  const { controls, getEditorState, onChange } = props;
  return (
    <div className="Draftail-Toolbar" role="toolbar">
      <ToolbarDefaults {...props} />

      <ToolbarGroup>
        {controls.map((control, i) => {
          if (control.inline || control.meta) {
            return null;
          }

          const Control = control.block || control;

          return (
            <Control
              // eslint-disable-next-line react/no-array-index-key
              key={i}
              getEditorState={getEditorState}
              onChange={onChange}
            />
          );
        })}
      </ToolbarGroup>
    </div>
  );
};

export default Toolbar;
