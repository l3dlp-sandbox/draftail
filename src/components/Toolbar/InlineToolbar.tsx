import React, { useContext, useMemo, useState } from "react";

import { Control } from "../../api/types";

import Toolbar, { ToolbarProps } from "./Toolbar";
import ToolbarButton from "./ToolbarButton";
import FloatingToolbar from "./FloatingToolbar/FloatingToolbar";

type ToolbarType = "floating" | "sticky";

interface PinButton extends Control {
  floatingLabel?: Control["label"];
  stickyLabel?: Control["label"];
  floatingDescription?: Control["description"];
  stickyDescription?: Control["description"];
  floatingIcon?: Control["icon"];
  stickyIcon?: Control["icon"];
}

const ToolbarContext = React.createContext<{
  pinButton: PinButton;
  toolbar: ToolbarType;
  setToolbar: (type: ToolbarType) => void;
}>({ pinButton: {}, toolbar: "sticky", setToolbar: () => {} });

const ToolbarPinButton = () => {
  const { pinButton, toolbar, setToolbar } = useContext(ToolbarContext);
  const label =
    toolbar === "floating"
      ? pinButton.floatingLabel || "↗"
      : pinButton.stickyLabel || "↙";
  const description =
    toolbar === "floating"
      ? pinButton.floatingDescription
      : pinButton.stickyDescription;
  const icon =
    toolbar === "floating" ? pinButton.floatingIcon : pinButton.stickyIcon;
  return (
    <ToolbarButton
      name={pinButton.type || "PIN_TOOLBAR"}
      className="Draftail-ToolbarButton--pin"
      tooltipDirection="down"
      title={description || pinButton.description}
      icon={icon || pinButton.icon}
      label={icon || pinButton.icon ? null : label}
      onClick={setToolbar.bind(
        null,
        toolbar === "floating" ? "sticky" : "floating",
      )}
      active={false}
    />
  );
};

const pinControls = [
  {
    inline: () => <ToolbarPinButton />,
  },
];

export interface InlineToolbarProps extends ToolbarProps {
  pinButton?: PinButton;
  defaultToolbar?: ToolbarType;
  onSetToolbar?: (toolbar: ToolbarType, callback: () => void) => void;
}

const InlineToolbar = ({
  pinButton = {},
  defaultToolbar = "floating",
  onSetToolbar,
  controls,
  ...otherProps
}: InlineToolbarProps) => {
  const [toolbar, setToolbar] = useState(defaultToolbar);
  const ctx = useMemo(
    () => ({
      pinButton,
      toolbar,
      setToolbar: (val: ToolbarType) => {
        if (onSetToolbar) {
          onSetToolbar(val, () => {
            setToolbar(val);
          });
        } else {
          setToolbar(val);
        }
      },
    }),
    [pinButton, toolbar, setToolbar, onSetToolbar],
  );
  const ToolbarComponent = toolbar === "floating" ? FloatingToolbar : Toolbar;

  return (
    <ToolbarContext.Provider value={ctx}>
      <ToolbarComponent
        controls={controls.concat(pinControls)}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...otherProps}
      />
    </ToolbarContext.Provider>
  );
};

export default InlineToolbar;
