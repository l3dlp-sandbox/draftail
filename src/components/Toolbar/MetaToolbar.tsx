import React from "react";

import { LegacyControlControl } from "../../api/types";
import { getControlLabel, showControl } from "../../api/ui";
import ToolbarButton from "./ToolbarButton";
import ToolbarGroup from "./ToolbarGroup";
import { getButtonTitle } from "./ToolbarDefaults";
import { ToolbarProps } from "./Toolbar";

export interface MetaToolbarProps extends ToolbarProps {
  showBlockEntities?: boolean;
}

const MetaToolbar = ({
  showBlockEntities,
  entityTypes,
  controls,
  getEditorState,
  onChange,
  onRequestSource,
}: MetaToolbarProps) => (
  <div className="Draftail-MetaToolbar">
    {showBlockEntities ? (
      <ToolbarGroup key="entities" name="entities">
        {entityTypes
          .filter((entityType) => showControl(entityType) && entityType.block)
          .map((t) => (
            <ToolbarButton
              key={t.type}
              name={t.type}
              onClick={onRequestSource}
              label={getControlLabel(t.type, t)}
              title={getButtonTitle(t.type, t)}
              icon={t.icon}
            />
          ))}
      </ToolbarGroup>
    ) : null}
    <ToolbarGroup name="controls">
      {controls.map((control, i) => {
        if (control.inline || control.block) {
          return null;
        }

        // Support the legacy and current controls APIs.
        const Control = control.meta || (control as LegacyControlControl);

        return (
          <Control
            // eslint-disable-next-line react/no-array-index-key
            key={i}
            getEditorState={getEditorState}
            onChange={onChange}
          />
        );
      })}
    </ToolbarGroup>
  </div>
);

export default MetaToolbar;
