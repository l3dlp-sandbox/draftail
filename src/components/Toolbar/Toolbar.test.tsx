import { OrderedSet } from "immutable";
import React from "react";
import { EditorState } from "draft-js";
import { shallow } from "enzyme";

import Toolbar from "./Toolbar";

const mockProps = {
  currentStyles: OrderedSet(),
  currentBlock: "unstyled",
  currentBlockKey: "abcd",
  enableHorizontalRule: false,
  enableLineBreak: false,
  showUndoControl: false,
  showRedoControl: false,
  entityTypes: [],
  blockTypes: [],
  inlineStyles: [],
  commands: false,
  toggleBlockType: () => {},
  toggleInlineStyle: () => {},
  addHR: () => {},
  addBR: () => {},
  onUndoRedo: () => {},
  onRequestSource: () => {},
  onCompleteSource: () => {},
  focus: () => {},
  getEditorState: () => EditorState.createEmpty(),
  onChange: () => {},
  controls: [],
};

describe("Toolbar", () => {
  it("empty", () => {
    expect(shallow(<Toolbar {...mockProps} />)).toMatchSnapshot();
  });

  it("#controls", () => {
    expect(
      shallow(
        <Toolbar
          {...mockProps}
          controls={[{ block: () => <span>Test</span> }]}
        />,
      ),
    ).toMatchSnapshot();
  });
});
