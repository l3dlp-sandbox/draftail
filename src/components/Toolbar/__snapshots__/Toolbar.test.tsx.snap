// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar #controls 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarDefaults
    addBR={[Function]}
    addHR={[Function]}
    blockTypes={Array []}
    commands={false}
    controls={
      Array [
        Object {
          "block": [Function],
        },
      ]
    }
    currentBlock="unstyled"
    currentBlockKey="abcd"
    currentStyles={Immutable.OrderedSet []}
    enableHorizontalRule={false}
    enableLineBreak={false}
    entityTypes={Array []}
    focus={[Function]}
    getEditorState={[Function]}
    inlineStyles={Array []}
    onChange={[Function]}
    onCompleteSource={[Function]}
    onRequestSource={[Function]}
    onUndoRedo={[Function]}
    showRedoControl={false}
    showUndoControl={false}
    toggleBlockType={[Function]}
    toggleInlineStyle={[Function]}
  />
  <ToolbarGroup>
    <block
      getEditorState={[Function]}
      key="0"
      onChange={[Function]}
    />
  </ToolbarGroup>
</div>
`;

exports[`Toolbar empty 1`] = `
<div
  className="Draftail-Toolbar"
  role="toolbar"
>
  <ToolbarDefaults
    addBR={[Function]}
    addHR={[Function]}
    blockTypes={Array []}
    commands={false}
    controls={Array []}
    currentBlock="unstyled"
    currentBlockKey="abcd"
    currentStyles={Immutable.OrderedSet []}
    enableHorizontalRule={false}
    enableLineBreak={false}
    entityTypes={Array []}
    focus={[Function]}
    getEditorState={[Function]}
    inlineStyles={Array []}
    onChange={[Function]}
    onCompleteSource={[Function]}
    onRequestSource={[Function]}
    onUndoRedo={[Function]}
    showRedoControl={false}
    showUndoControl={false}
    toggleBlockType={[Function]}
    toggleInlineStyle={[Function]}
  />
  <ToolbarGroup />
</div>
`;
